File: leet/0023_merge_k_sorted_lists/0023_merge_k_sorted_lists.js
// LeetCode 23. Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/
// O(nlogk) time, O(logk) space
var mergeKLists = function(lists) {
    
    let divideAndConquerLists = function(lists, start, end) {
        if (start >= end)
            return lists[start];
        let mid = parseInt(start + (end - start) / 2);
        let left = divideAndConquerLists(lists, start, mid);
        let right = divideAndConquerLists(lists, mid + 1, end);
        return mergeTwoLists(left, right);
    }
    
    let mergeTwoLists = function(list1, list2) {
        let dummy = new ListNode(), head = dummy;
        while (list1 && list2) {
            if (list1.val <= list2.val) {
                head.next = list1;
                list1 = list1.next;
            } else {
                head.next = list2;
                list2 = list2.next;
            }
            head = head.next;
        }
        if (list1)
            head.next = list1;
        if (list2)
            head.next = list2;
        return dummy.next;
    }
    
    if (!lists.length)
        return null;
    return divideAndConquerLists(lists, 0, lists.length);
};

File: leet/0023_merge_k_sorted_lists/0023_merge_k_sorted_lists.py
# LeetCode 23. Merge k Sorted Lists
# https://leetcode.com/problems/merge-k-sorted-lists/
# O(nlogk) time, O(k) space
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        saveHead = head = ListNode(0)
        heap = [(node.val, i, node) for i, node in enumerate(lists) if node]
        heapq.heapify(heap)
        while heap:
            val, i, node = heapq.heappop(heap)
            head.next = node
            head = head.next
            if node.next:
                heapq.heappush(heap, (node.next.val, i, node.next))
        return saveHead.next
